Make `but_api::legacy::diff::commit_details()` non-legacy with macro-support.

Byron opened the pull request:
This PR will add the machinery that's needed to be able to port legacy functions that take `ProjectId` into the modern world, so they take `Context` instead.
The main difficulty I see is to make `api_cmd` deal with the conversion of `tauri` and the server, down to either `&Context`, or `&mut Context`.

Let's see what kind of simplifications happen on the way, as `Context` is also doable, at the expense of automated caching,
i.e. re-use of the context for a particular project which would require the `Context` to be passed as reference.

### Tasks

* [x] manual port of `commit_details()`
    - [x] make sure line stats aren't computed unnecessarily
    - [x] move CommitDetails into `but-core`
* [x] `api_cmd` update to be able to generate the tauri and server versions.
    - This will get us the 'cmd' version, the one that takes the params struct for the but-server implementation
* [x] let UI use the new version of the function, remove the old one and remove the `_v2` suffix.
* [x] address PR review
    - [x] auto ident rename when mapping
    - [x] update but_api docs
    - [x] double-check `try_from` error handling

Byron committed:
Make `but_api::legacy::diff::commit_details()` non-legacy.

Byron committed:
Refactor the api-macro to prepare it for improvements.

Byron committed:
Add support for converting select types automatically by code injection.

The idea is that for the types we know, we can implement a type conversion
from the pure API type to one that can be deserialised.
And from there, these types can be converted back into their API types
before being passed to their API counterparts.

Byron committed:
Remove the legacy version of `commit_details` in favor of the modern version.

Adapt the frontend as well so it's able to get line statistics.

Byron committed:
Rename the macro to the absolute epic... `but_api` :).

It seems to fit much better than `api_cmd` or `api_cmd_tauri`,
as it literally is what makes our API, at least the compatibility
layers.

Byron committed:
Address Copilot review comments

Byron commented on crates/but-api/src/diff.rs:
```diff
@@ -0,0 +1,75 @@
+    project_id: &Context,
+    commit_id: gix::ObjectId,
+) -> anyhow::Result<CommitDetails> {
+    commit_details(project_id, commit_id, ComputeLineStats::Yes)
```
Intentional

Byron commented on crates/but-api-macros/src/lib.rs:
```diff
@@ -19,202 +20,439 @@ pub fn api_cmd_tauri(attr: TokenStream, item: TokenStream) -> TokenStream {
+                        }
+                        FromMode::TryFrom => {
+                            quote! {
+                                let mut #name = <#ty>::try_from(#name)?;
```
It's a pre-emptive `mut`, intended. Without it, `&mut` wouldn't be supported.

krlvi commented on crates/but-api/src/diff.rs:
```diff
@@ -0,0 +1,74 @@
+#[but_api(json::CommitDetails)]
+#[instrument(err(Debug))]
+pub fn commit_details_with_line_stats(
+    project_id: &Context,
```
Oh I know why - in order to not change the frontend. And also to not leak `ctx` as a name into the frontend.

Byron replied:
Yes, that's true.

It's a crutch and it comes at a cost, but I can see the light… *<after fending off the heart attack by pounding the chest a couple of times, the light disappears and I continue writing>* and … actually there will always be this difference between the UI passing `projectId` or `projectHandle`, which has to get transformed into `ctx`.

To my mind the problem is mostly that the frontend has to rely on parameter names.
Maybe this is just something we document and live with? Maybe it's something to change to `projectHandle` or something shorter as well, or maybe `projectId` is fine even if the type eventually becomes the self-describing `ProjectHandle`.

In the end, I also don't see any other solution but to document that clearly, for those frontend implementations that rely on parameter names at least. Dioxus for example wouldn't have that problem at all and calls the Rust API directly.

krlvi commented on crates/but-ctx/src/legacy.rs:
```diff
@@ -63,6 +63,15 @@ impl Context {
+
+    fn try_from(value: LegacyProjectId) -> Result<Self, Self::Error> {
+        let project = gitbutler_project::get(value)?;
+        Context::new_from_legacy_project(project)
```
Since this gets executed from within a macro... in theory error could be extra confusing. I wonder, perhaps this is a place we could be helpful when debugging if we included some additional error info with the anyhow's `.context`? What do you think?

krlvi reviewed (approved):
I left some reflections/ comments. Curious about the E2E test failure but other than that looks good

Byron committed:
Address review by Kiril

 auto ident rename when mapping
 update but_api docs
 double-check try_from error handling

Byron commented:
For the E2E failure, the crucial lesson learned here that one must not only search for `'commit_details'`, but also for `"commit_details"` as playwright uses the `but-server` as backend. Makes perfect sense as well, and I hope to remember it for a while now.

kritsada1998mt-create commented:
[]()
